File: dms-frontend/src/app/core/services/document.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { ApiResponse } from '../../shared/models/api-response.model';
import { DocumentResponse } from '../../shared/models/document.model';
import { PagedResponse } from '../../shared/models/paged-response.model';

@Injectable({ providedIn: 'root' })
export class DocumentService {
    private readonly apiUrl = `${environment.apiUrl}/documents`;
    constructor(private readonly http: HttpClient) { }

    uploadDocument(file: File, title: string): Observable<ApiResponse<DocumentResponse>> {
        const formData = new FormData();
        formData.append('file', file);
        formData.append('title', title);
        return this.http.post<ApiResponse<any>>(`${this.apiUrl}/upload`, formData);
    }

    getAllDocuments(): Observable<ApiResponse<DocumentResponse[]>> {
        return this.http.get<ApiResponse<DocumentResponse[]>>(this.apiUrl);
    }

    getPagedDocuments(page: number = 0, size: number = 10, statuses?: string[]): Observable<ApiResponse<PagedResponse<DocumentResponse>>> {
        let url = `${this.apiUrl}/paged?page=${page}&size=${size}`;
        if (statuses && statuses.length > 0) {
            url += `&statuses=${statuses.join(',')}`;
        }
        return this.http.get<ApiResponse<PagedResponse<DocumentResponse>>>(url);
    }

    getDocumentById(id: number): Observable<ApiResponse<DocumentResponse>> {
        return this.http.get<ApiResponse<DocumentResponse>>(`${this.apiUrl}/${id}`);
    }

    downloadDocument(id: number): Observable<Blob> {
        return this.http.get(`${this.apiUrl}/${id}/download`, { responseType: 'blob' });
    }
}

File: dms/src/main/java/com/example/documentmanagement/service/DocumentService.java
package com.example.documentmanagement.service;

import com.example.documentmanagement.dto.response.DocumentResponse;
import com.example.documentmanagement.dto.response.PagedResponse;
import org.springframework.core.io.Resource;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.security.NoSuchAlgorithmException;
import java.util.List;

public interface DocumentService {

    DocumentResponse uploadDocument(MultipartFile file, String title) throws IOException, NoSuchAlgorithmException;

    DocumentResponse getDocumentById(Long id);

    List<DocumentResponse> getAllDocuments();

    PagedResponse<DocumentResponse> getPagedDocuments(int page, int size, List<String> statuses);

    Resource downloadDocument(Long id);
}
